jwql/shared_tasks/run_pipeline.py

```diff
@@ -43,27 +44,65 @@ def run_pipe(input_file, short_name, work_directory, instrument, outputs, max_co
     status_file_name = short_name + "_status.txt"
     status_file = os.path.join(work_directory, status_file_name)
     uncal_file = os.path.join(work_directory, input_file_basename)
-    
+
     with open(status_file, 'a+') as status_f:
         status_f.write("Running run_pipe\n")
         status_f.write("\t input_file_basename is {} ({})\n".format(input_file_basename, type(input_file_basename)))
         status_f.write("\t start_dir is {} ({})\n".format(start_dir, type(start_dir)))
         status_f.write("\t uncal_file is {} ({})\n".format(uncal_file, type(uncal_file)))
-    
+        status_f.write(f"\t outputs is {outputs}\n")
+
     try:
         copy_files([input_file], work_directory)
         set_permissions(uncal_file)
-    
+
         steps = get_pipeline_steps(instrument)
+
+        # If the input file is a file other than uncal.fits, then we may only need to run a
+        # subset of steps. Check the completed steps in the input file. Find the latest step
+        # that has been completed, and skip that plus all prior steps
+        if 'uncal.fits' not in input_file:
+            completed_steps = completed_pipeline_steps(input_file)
+
+            # Reverse the boolean value, so that now steps answers the question: "Do we need
+            # to run this step?""
+            for step in steps:
+                steps[step] = not completed_steps[step]
+
+        # Make sure we don't run steps out of order. Find the latest step that has been
+        # run, and only run subsequent steps. This protects against cases where some early
+        # step was not run. In that case, we don't want to go back and run it because running
+        # pipeline steps out of order doesn't work.
+        last_run = 'group_scale'  # initialize to the first step
+        for step in steps:
+            if not steps[step]:
+                last_run = deepcopy(step)
+
+        for step in steps:
+            if step == last_run:
+                break
+            if step != last_run:
+                steps[step] = False
+
+        # Set any steps the user specifically asks to skip
+        for step, step_dict in step_args.items():
+            if 'skip' in step_dict:
+                print(f'SKIP the {step}!!')
+                if step_dict['skip']:
+                    steps[step] = False
+
+        # Run each specified step
         first_step_to_be_run = True
         for step_name in steps:
-            sys.stderr.write("Running step {}\n".format(step_name))
```

Comment:
@york-stsci does it make sense to add an entry about running a pipeline step to stderr?

Reply:
I'm not sure it needs to be a write to stderr, but it should write using one of STDOUT or STDERR, because the shared tasks function(s) get log information from the subprocess by redirecting STDOUT, STDERR, or both to a pipe, and then redirecting that pipe to a log (all in the shared_tasks.py file).

Reply:
Okay, looking at it, in the file "shared_tasks.py",

- At line 202, function `log_subprocess_output()` adds to the log file anything that comes out of the pipe produced by calling `popen()` on the subprocess
- In `run_subprocess()`, lines 235-237, when calling `subprocess.Popen()`, the standard error output from the subprocess is sent to pipe, and `log_subprocess_output()` is called on `process.stderr`

Now, the reason that I'm doing this is to get the pipeline logging, because the JWST pipeline prints all of its output, whatever the source, to STDERR rather than STDOUT, presumably for *some* reason, but I don't know what the reason actually *is*.

Reply:
Thanks @york-stsci so for logging purposes should we go ahead and add the line back in?

Reply:
Ahh nevermind, I just took a look at the file. Looks like the line is currently there from the most recent changes.

Reply:
Yes, I would say leave it in.